### PR TITLE
Introduce a filtered collector manager

### PR DESCRIPTION
**Pull request by javanna**
In order to add support for inter-segment search concurrency, we need to implement collector managers for all of our custom collectors.

This PR introduces a collector manager that is based on `FilteredCollector`, used when a post_filter is provided as part of a search request.

Note that the collector manager is not yet integrated in the query phase.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/common/lucene/search/FilteredCollectorTests.java [hunk 1]
```diff
@@ -39,7 +41,7 @@ public void setUp() throws Exception {
         super.setUp();
         directory = newDirectory();
         RandomIndexWriter writer = new RandomIndexWriter(random(), directory, newIndexWriterConfig());
-        numDocs = randomIntBetween(10, 100);
+        numDocs = randomIntBetween(900, 1000);
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/common/lucene/search/FilteredCollectorTests.java:44]: This is to increase the likelihood of leveraging concurrency. It still does not happen super often, around 10% of the runs. Adding flushes would help but random index writer already does it and it also slows down the tests. I have opened [12369](https://github.com/apache/lucene/pull/12369) to potentially address this in Lucene,

**Comment by javanna** [comment]
run elasticsearch-ci/part-1

**Review by cbuescher — APPROVED** [review]
Thanks, I think this is a good template for creating other managers as well. LGTM

File: server/src/test/java/org/elasticsearch/common/lucene/search/FilteredCollectorTests.java (hunk 1 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/common/lucene/search/FilteredCollectorTests.java:44]: The creation of segments and slices should be fully reproducible with the seed, if it's now it sounds like a bug.

**Pull request merged**